synapse/handlers/room_member.py

```diff
@@ -178,6 +179,59 @@ async def ratelimit_invite(
 
         await self._invites_per_user_limiter.ratelimit(requester, invitee_user_id)
 
+    async def _can_join_restricted_room(
+        self, state_ids: StateMap[str], room_version: RoomVersion, user_id: str
+    ) -> bool:
+        """
+        Check whether a user can join a restricted room.
```

Comment:
this gets run on *all* rooms, not just restricted ones, so I think it needs a rename and a wording tweak... maybe `_can_join_without_invite` ?